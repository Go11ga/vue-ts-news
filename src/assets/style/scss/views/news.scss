.vtn-news {
    height: 100%;
    
    margin: 0 -10px;

    display: flex;
    justify-content: space-between;
    align-content: flex-start;
    flex-wrap: wrap;

    &__empty {
        height: 100%;
        width: 100%;
        padding: 100px 0;

        display: flex;
        justify-content: center;
        align-items: center;

        font-size: 24px;
        font-weight: 700;
    }
}
