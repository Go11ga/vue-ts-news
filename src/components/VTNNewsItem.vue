<template>
    <div
        class="vtn-newsitem"
        :class="{ 'vtn-newsitem--list' : news.viewMode === 'list' }"
    >
        <div class="vtn-newsitem__inner">

            <div class="vtn-newsitem__header">
                <a class="vtn-newsitem__title" :href="newsItem.url"  target="_blank">
                    {{ newsItem.title }}
                </a>
            </div>

            <div class="vtn-newsitem__content">
                <img v-if="newsItem.urlToImage" class="vtn-newsitem__img" :src="newsItem.urlToImage" >
                <img v-else class="vtn-newsitem__img" src="https://dummyimage.com/300x168/DCDCDC/fff">

                <div
                    class="vtn-newsitem__text"
                    :class="{'vtn-newsitem__text--height-auto' : heightAuto}"
                >
                    {{ newsItem.description }}
                </div>

                <button
                    v-if="newsItem.description?.length" 
                    class="vtn-newsitem__btn"
                    @click="toggleHeight"
                >
                    Подробнее
                </button>
            </div>

            <div class="vtn-newsitem__footer">
                <a class="vtn-newsitem__channel-link" :href="getLink(newsItem.url)" target="_blank">
                    {{ getLink(newsItem.url) }}
                </a>

                <div class="vtn-newsitem__pubdate">
                    {{ getDate(newsItem.publishedAt) }}
                </div>
            </div>

        </div>
    </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { INewsItem } from '@/types/interfaces'
import { mapState } from 'vuex'

export default defineComponent({
    name: 'VTNNewsItem',

    props: {
        newsItem: {
            type: Object as PropType<INewsItem>,
            required: true
        }
    },

    data() {
        return {
            heightAuto: false
        }
    },

    computed: {
        ...mapState([
            'news'
        ])
    },

    methods: {
        getLink(url: string): string {
            let strs = url.split('/')

            return `https://${strs[2]}`
        },

        getDate(date: string): string {
            let time = new Date(date)

            return `${time.toLocaleDateString()} ${time.toLocaleTimeString()}`
        },

        toggleHeight() {
            this.heightAuto = !this.heightAuto
        }
    }
})
</script>
