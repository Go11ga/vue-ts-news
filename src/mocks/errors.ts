import { IError, IStateErrors } from '@/types/interfaces'

const makeState = (error: IError = {}): IStateErrors => {
    return {
        error
    }
}

export { makeState }
