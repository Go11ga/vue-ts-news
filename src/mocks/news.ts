import { INewsItem, IStateNews } from '@/types/interfaces'
import { AxiosError } from 'axios'

const newsItem_01: INewsItem = {
    author: 'Author_01',
    content: 'Content_01',
    description: 'Description_01',
    publishedAt: '2023-01-03T20:48:00Z',
    source: {
        id: null,
        name: "Sorce_01"
    },
    title: 'Title_01',
    url: 'https://domain_01.com/domain_01',
    urlToImage: 'https://domain_01_img.com'
}

const newsItem_02: INewsItem = {
    author: 'Author_02',
    content: 'Content_02',
    description: null,
    publishedAt: '2023-01-05T20:48:00Z',
    source: {
        id: null,
        name: "Sorce_02"
    },
    title: 'Title_02',
    url: 'https://domain_02.com/domain_02',
    urlToImage: null
}

const axiosError_426 = {
    config: {},
    request: {},
    response: {
        status: 426,
        data: {
            message: 'axiosError_426'
        }
    }
} as AxiosError<any>

const axiosError_500 = {
    config: {},
    request: {},
    response: {},
    message: 'Test error message'
} as AxiosError<any>

const makeState = (activeChannel: string, searchValue: string, news: Array<INewsItem> = []): IStateNews => {
    return {
        channels: [
            { id: 1, code: 'all', title: 'Все' },
            { id: 2, code: 'lenta', title: 'Lenta' },
            { id: 3, code: 'bbc-news', title: 'BBC' }
        ],
        activeChannel,
        viewMode: '',
        news,
        totalResults: 1,
        currentPage: 1,
        newsPerPage: 20,
        newsProcess: false,
        searchValue
    }
}

export { newsItem_01, newsItem_02, axiosError_426, axiosError_500, makeState }