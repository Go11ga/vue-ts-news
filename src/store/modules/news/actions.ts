import { GetterTree, ActionTree, Commit } from 'vuex'
import { IStateRoot, IStateNews, INewsResponse } from '@/types/interfaces'
import axios from 'axios'

const actions: ActionTree<IStateNews, IStateRoot> = {
    async fetchNews({ state, getters, commit } : { state: IStateNews, getters: GetterTree<IStateNews, IStateRoot>, commit: Commit }) {
        try {
            state.newsProcess = true

            const { data } = await axios.get<INewsResponse>(`https://newsapi.org/v2/everything?${getters.getQuery}&apiKey=${process.env.VUE_APP_KEY}`)
            
            state.news = data.articles
            state.totalResults = data.totalResults

        } catch (error) {
            if (axios.isAxiosError(error)) {
                if (error?.response?.status === 426 || error?.response?.status === 429) {
                    commit('SET_ERROR', { type: 'error', msg: error?.response?.data?.message })
                    return
                }

                commit('SET_ERROR', { type: 'error', msg: error.message })

            } else {
                commit('SET_ERROR', { type: 'error', msg: 'An unexpected error occurred' })
            }
        } finally {
            state.newsProcess = false
        }
    }
}

export default actions
