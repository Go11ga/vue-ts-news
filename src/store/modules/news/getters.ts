import { GetterTree } from 'vuex';
import { IStateRoot, IStateNews } from '@/types/interfaces'

const getters: GetterTree<IStateNews, IStateRoot> = {
    getSources(state: IStateNews) {
        const arr = [...state.channels].splice(1, state.channels.length)

        let str = ''
        arr.forEach(el => str += `${el.code},`)
        str = str.slice(0, -1)

        return state.activeChannel === 'all' ? str : state.activeChannel
    },

    getSearch(state: IStateNews) {
        return state.searchValue.length > 0 ? `q="${state.searchValue}"` : `q`
    },

    getQuery(state: IStateNews, getters: GetterTree<IStateNews, IStateRoot>) {
        return `sources=${getters.getSources}&sortBy=publishedAt&pageSize=${state.newsPerPage}&page=${state.currentPage}&${getters.getSearch}`
    }
}

export default getters
