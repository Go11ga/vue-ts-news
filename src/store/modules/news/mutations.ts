import { MutationTree } from 'vuex'
import { IStateNews } from '@/types/interfaces'

const mutation: MutationTree<IStateNews> = {
    SET_ACTIVE_CHANNEL(state: IStateNews, payload: string) {
        state.activeChannel = payload  
    },

    SET_VIEW_MODE(state: IStateNews, payload: string) {
        state.viewMode = payload 
    },

    SET_CURRENT_PAGE(state: IStateNews, payload: number) {
        state.currentPage = payload
    },

    CLEAR_NEWS(state: IStateNews) {
        state.news.length = 0
    },

    SET_SEARCH_VALUE(state: IStateNews, payload: string) {
        state.searchValue = payload
    }
}

export default mutation
