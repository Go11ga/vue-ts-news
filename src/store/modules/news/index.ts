import { Module } from 'vuex';
import { IStateRoot, IStateNews } from '@/types/interfaces'

import state from '@/store/modules/news/state'
import getters from '@/store/modules/news/getters'
import mutations from '@/store/modules/news/mutations'
import actions from '@/store/modules/news/actions'

const newsModule: Module<IStateNews, IStateRoot> = {
    state,
    getters,
    mutations,
    actions
}

export { newsModule }
