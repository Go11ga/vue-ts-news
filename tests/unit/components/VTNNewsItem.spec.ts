import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import VTNNewsItem from '@/components/VTNNewsItem.vue'
import { newsItem_01, newsItem_02 } from '@/mocks/news'

const makeWrapper = (viewMode: string = 'list') => {
    const state = () => {
        return {
            news: {
                viewMode
            }
        }
    }

    const store = createStore({
        state
    })

    const wrapper = shallowMount(VTNNewsItem, {
        props: {
            newsItem: newsItem_01
        },
        global: {
            plugins: [store]
        }
    })

    return wrapper
}

describe('VTNNewsItem', () => {
    let wrapper: any

    beforeEach(() => {
        wrapper = makeWrapper()
    })

    afterEach(() => {
        wrapper.unmount()
    })

    it('dummy name test', () => {
        expect(wrapper.vm.$options.name).toMatch('VTNNewsItem')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('отрисовка и пропсы', async () => {
        expect(wrapper.find('.vtn-newsitem__title').text()).toMatch('Title_01')
        expect(wrapper.find('.vtn-newsitem__title').attributes().href).toMatch('https://domain_01.com/domain_01')
        expect(wrapper.find('.vtn-newsitem__img').attributes().src).toMatch('https://domain_01_img.com')
        expect(wrapper.find('.vtn-newsitem__text').text()).toMatch('Description_01')
        expect(wrapper.html().includes('vtn-newsitem__btn')).toBe(true)
        expect(wrapper.find('.vtn-newsitem__channel-link').attributes().href).toMatch('https://domain_01.com')
        expect(wrapper.find('.vtn-newsitem__channel-link').text()).toMatch('https://domain_01.com')
        expect(wrapper.find('.vtn-newsitem__pubdate').text()).toMatch('03.01.2023 23:48:00')

        wrapper.setProps({
            newsItem: newsItem_02
        })

        await wrapper.vm.$nextTick()

        expect(wrapper.find('.vtn-newsitem__title').text()).toMatch('Title_02')
        expect(wrapper.find('.vtn-newsitem__title').attributes().href).toMatch('https://domain_02.com/domain_02')
        expect(wrapper.find('.vtn-newsitem__img').attributes().src).toMatch('https://dummyimage.com/300x168/DCDCDC/fff')
        expect(wrapper.find('.vtn-newsitem__text').text()).toMatch('')
        expect(wrapper.html().includes('vtn-newsitem__btn')).toBe(false)
        expect(wrapper.find('.vtn-newsitem__channel-link').attributes().href).toMatch('https://domain_02.com')
        expect(wrapper.find('.vtn-newsitem__channel-link').text()).toMatch('https://domain_02.com')
        expect(wrapper.find('.vtn-newsitem__pubdate').text()).toMatch('05.01.2023 23:48:00')
    })

    it('изменение state viewMode', async () => {
        expect(wrapper.find('.vtn-newsitem').classes()).toContain('vtn-newsitem--list')

        wrapper = makeWrapper('grid')

        await wrapper.vm.$nextTick()

        expect(wrapper.find('.vtn-newsitem').classes()).not.toContain('vtn-newsitem--list')
    })

    it('метод getLink', () => {
        let res1 = wrapper.vm.getLink('https://domain_01.com/domain_01')
        expect(res1).toMatch('https://domain_01.com')

        let res2 = wrapper.vm.getLink('')
        expect(res2).toMatch('https://undefined') 
    })

    it('метод getDate', () => {
        let res1 = wrapper.vm.getDate('2023-01-03T20:48:00Z')
        expect(res1).toMatch('03.01.2023 23:48:00')

        let res2 = wrapper.vm.getDate('')
        expect(res2).toMatch('Invalid Date Invalid Date') 
    })

    it('клик по кнопке Подробнее вызывает метод toggleHeight', () => {
        wrapper.vm.toggleHeight = jest.fn()

        wrapper.find('.vtn-newsitem__btn').trigger('click')
        expect(wrapper.vm.toggleHeight).toHaveBeenCalled()
    })

    it('метод toggleHeight', async () => {
        wrapper.vm.toggleHeight()

        expect(wrapper.vm.heightAuto).toBe(true)
        await wrapper.vm.$nextTick()
        expect(wrapper.find('.vtn-newsitem__text').classes()).toContain('vtn-newsitem__text--height-auto')

        wrapper.vm.toggleHeight()

        expect(wrapper.vm.heightAuto).toBe(false)
        await wrapper.vm.$nextTick()
        expect(wrapper.find('.vtn-newsitem__text').classes()).not.toContain('vtn-newsitem__text--height-auto')
    })
})
