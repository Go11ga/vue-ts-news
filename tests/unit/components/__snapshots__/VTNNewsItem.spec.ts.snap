// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTNNewsItem snapshot 1`] = `
<div
  class="vtn-newsitem vtn-newsitem--list"
>
  <div
    class="vtn-newsitem__inner"
  >
    <div
      class="vtn-newsitem__header"
    >
      <a
        class="vtn-newsitem__title"
        href="https://domain_01.com/domain_01"
        target="_blank"
      >
        Title_01
      </a>
    </div>
    <div
      class="vtn-newsitem__content"
    >
      <img
        class="vtn-newsitem__img"
        src="https://domain_01_img.com"
      />
      <div
        class="vtn-newsitem__text"
      >
        Description_01
      </div>
      <button
        class="vtn-newsitem__btn"
      >
         Подробнее 
      </button>
    </div>
    <div
      class="vtn-newsitem__footer"
    >
      <a
        class="vtn-newsitem__channel-link"
        href="https://domain_01.com"
        target="_blank"
      >
        https://domain_01.com
      </a>
      <div
        class="vtn-newsitem__pubdate"
      >
        03.01.2023 23:48:00
      </div>
    </div>
  </div>
</div>
`;
